processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,405 @@
+        }
+        if (now >= spinTimeoutAt) {
+          Thread.yield();
+        }
+        now = System.nanoTime();
+      }
+      curWriteIndex = 0;
+    } else {
+      final int nextWriteIndex = curWriteIndex + 1;
+      while ((nextWriteIndex == nextReadIndex) && !Thread.currentThread().isInterrupted()) {
+        if (now >= queryTimeoutAtNs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+        if (now >= spinTimeoutAt) {
+          Thread.yield();
+        }
+        now = System.nanoTime();
+      }
+      curWriteIndex = nextWriteIndex;
+    }
+  }
+
+  @Override
+  public void reset()
+  {
+    curWriteIndex = -1;
+    nextReadIndex = -1;
+    finished = false;
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  /**
+   * Signal that no more inputs are added.  Must be called after {@link #aggregate(Object)} is called for the last input.
+   */
+  public void finish()
+  {
+    increaseWriteIndex();
+    finished = true;
+  }
+
+  /**
+   * Return a sorted iterator.  This method can be called safely while writing and iterating thread and writing thread
+   * can be different.  The result iterator always returns sorted results.  This method should be called only one time
+   * per grouper.
+   *
+   * @return a sorted iterator
+   */
+  public CloseableIterator<Entry<KeyType>> iterator()
+  {
+    if (!initialized) {
+      throw new ISE("Grouper should be initialized first");
+    }
+
+    return new CloseableIterator<Entry<KeyType>>()
+    {
+      {
+        // Waits for some data to be ready
+        increaseReadIndex(0);
+      }
+
+      @Override
+      public boolean hasNext()
+      {
+        return !finished || remaining() > 0;
+      }
+
+      private int remaining()
+      {
+        if (curWriteIndex >= nextReadIndex) {
+          return curWriteIndex - nextReadIndex;
+        } else {
+          return (maxNumSlots - nextReadIndex) + curWriteIndex;
+        }
+      }
+
+      @Override
+      public Entry<KeyType> next()
+      {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+
+        final int recordOffset = recordSize * nextReadIndex;
+        final KeyType key = keySerde.fromByteBuffer(buffer, recordOffset);
+
+        final Object[] values = new Object[aggregators.length];
+        for (int i = 0; i < aggregators.length; i++) {
+          values[i] = aggregators[i].get(buffer, recordOffset + aggregatorOffsets[i]);
+        }
+
+        final int increaseTo = nextReadIndex == maxNumSlots - 1 ? 0 : nextReadIndex + 1;
+        increaseReadIndex(increaseTo);
+
+        return new Entry<>(key, values);
+      }
+
+      private void increaseReadIndex(int increaseTo)
```

Comment:
To me "increaseTo" name sounds like it's increment, not the value itself. Maybe think about better name.

Also please add comments to this class

Reply:
Changed to ```increaseReadIndexTo(int target)```.
I'll add comments soon.

Reply:
@leventov added comments. If you think more comments are needed, please let me know.